clusterloader2/framework/util/util.go

```diff
@@ -0,0 +1,154 @@
+
+package util
+
+import (
+	"fmt"
+	"time"
+
+	apiv1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+const (
+	// Parameters for retrying with exponential backoff.
+	retryBackoffInitialDuration = 100 * time.Millisecond
+	retryBackoffFactor          = 3
+	retryBackoffJitter          = 0
+	retryBackoffSteps           = 6
+)
+
+// RetryWithExponentialBackOff a utility for retrying the given function with exponential backoff.
+func RetryWithExponentialBackOff(fn wait.ConditionFunc) error {
+	backoff := wait.Backoff{
+		Duration: retryBackoffInitialDuration,
+		Factor:   retryBackoffFactor,
+		Jitter:   retryBackoffJitter,
+		Steps:    retryBackoffSteps,
+	}
+	return wait.ExponentialBackoff(backoff, fn)
+}
+
+// IsRetryableAPIError verifies whether the error is retryable.
+func IsRetryableAPIError(err error) bool {
+	// These errors may indicate a transient error that we can retry in tests.
+	if apierrs.IsInternalError(err) || apierrs.IsTimeout(err) || apierrs.IsServerTimeout(err) ||
+		apierrs.IsTooManyRequests(err) || utilnet.IsProbableEOF(err) || utilnet.IsConnectionReset(err) {
+		return true
+	}
+	// If the error sends the Retry-After header, we respect it as an explicit confirmation we should retry.
+	if _, shouldRetry := apierrs.SuggestsClientDelay(err); shouldRetry {
+		return true
+	}
+	return false
+}
+
+func retryOperation(operation func() error, isAllowedError func(error) bool) wait.ConditionFunc {
+	return func() (bool, error) {
+		err := operation()
+		if err == nil || (isAllowedError != nil && isAllowedError(err)) {
+			return true, nil
+		}
+		if IsRetryableAPIError(err) {
+			return false, nil
+		}
+		return false, err
+	}
+}
+
+// CreateNamespace creates a single namespace with given name.
+func CreateNamespace(c clientset.Interface, namespace string) error {
+	createFunc := func() error {
+		_, err := c.CoreV1().Namespaces().Create(&apiv1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}})
+		return err
+	}
+	return RetryWithExponentialBackOff(retryOperation(createFunc, apierrs.IsAlreadyExists))
+
+}
+
+// DeleteNamespace deletes namespace with given name.
+func DeleteNamespace(c clientset.Interface, namespace string) error {
+	deleteFunc := func() error {
+		return c.CoreV1().Namespaces().Delete(namespace, nil)
+	}
+	return RetryWithExponentialBackOff(retryOperation(deleteFunc, apierrs.IsNotFound))
+}
+
+// CreateObject creates object based on given object description.
+func CreateObject(dynamicClient dynamic.Interface, gvk schema.GroupVersionKind, namespace string, name string, obj *unstructured.Unstructured) error {
+	gvr, _ := meta.UnsafeGuessKindToResource(gvk)
```

Comment:
I don't like it - the whole purpose of introducing this ObjectType in the api was to avoid doing any conversions and tricks to determine params that you need to pass in dynamic client.
So if that's not exactly what you need, we should change the API to be exactly what you need.

That said, I took a bit of deeper look whether we actually need that. And I'm now pretty strongly convinced that we don't need this at all.
Instead, we should get this information from the object itself. This seems exactly like a function that you would like to copy. So please create a PR to remove the ObjectType type from api completely and start getting that information from there.

Reply:
TBH, I'm not sure if there is other way to do this (, assuming that we are using dynamic client). To do any operation on the cluster, first you need to call Resource(gvr), which is using resource instead of kind. Because only kind is provided by the object, conversion needs to be made.

Reply:
Once we start relying on objects for that, maybe that's needed - but that's fine.

I just didn't want to introduce a dedicate type, that doesn't really buy us anything, because we need to convert it twice anyway.